[RFC] Pull for automatic compilation database finding and better directory control

stsquad opened the pull request:
This fixes my issues with QEMU's build structure (where the compiler runs in a different build directory than the source file, see #1 ).  I'm making this an RFC pull request as I'm sure you'll want to make some comments on the way I've hacked this up ;-)

I'm not overly happy with the name clangcheck-build-path as build and compile db location could be two different things.

stsquad committed:
Add new option flycheck-clangcheck-dbname

This is unlikely to be tweaked but it will help with other cleanups to
put this in a var instead of copying and paste.

Signed-off-by: Alex Bennée <alex.bennee@linaro.org>

---
v2
  - document variable in README.org

stsquad committed:
Make the calculation of the build path more automagic

Given the compilation database generally sits at the root of your build
we can automate the process of looking for it. To avoid too much over
calculation I've made the variable local when set.

Additionally I've re-factored the code to extract bits from the JSON so
we can set `default-directory` before the checker is called. This should
ensure the build takes place in exactly the place the database says it
should.

This has removed the warning about the build directory being set but no
compilation database existing. This can be fixed but I suspect a bit
more re-factoring would be required.

Signed-off-by: Alex Bennée <alex.bennee@linaro.org>

kumar8600 commented:
looks good. thx for your contributions. but you should explain about new configuration variable `flycheck-clangcheck-dbname` in the README.org

stsquad committed:
Better handling of missing json case

stsquad committed:
Add a filter stage to getting the compile commands

There are some cases we don't won't blindly pass parameters to the
flycheck compile. One of these is dependency output as this will likely
pollute the real build when it runs.

---
v2:
  - use emacs-25/elpa seq-filter function
  - filter more dep generating options

kumar8600 commented:
Great. Sorry for the my delay. Thanks for your contributions.
